packages/node/src/indexer/mmr.service.ts

```diff
@@ -0,0 +1,149 @@
+      where: { key: 'blockOffset' },
+    });
+    if (blockOffset === null) {
+      throw new Error(`Poi service failed to fetch block offset from metadata`);
+    }
+    return Number(blockOffset.value);
+  }
+
+  async appendMmrNode(poiBlock: ProofOfIndex): Promise<void> {
+    let mmrRoot: Uint8Array;
+    const newLeaf = poiBlock.hash;
+    if (newLeaf.length !== DEFAULT_WORD_SIZE) {
+      throw new Error(
+        `Append Mmr failed, input data length should be ${DEFAULT_WORD_SIZE}`,
+      );
+    }
+    const blockLeafHeight = poiBlock.id - this.blockOffset - 1; //leaf index alway -1
+    const mmrNextLeafIndex = await this.fileBasedMmr.getLeafLength();
+    if (blockLeafHeight === mmrNextLeafIndex) {
+      //next mmr append leaf index should be same as block height
+      await this.fileBasedMmr.append(newLeaf, blockLeafHeight);
+      mmrRoot = await this.fileBasedMmr.getRoot(blockLeafHeight);
+    } else if (blockLeafHeight < mmrNextLeafIndex) {
+      //when FileBased Mmr height is ahead of Poi table, fetch mmr from fd and compare to poi
+      mmrRoot = await this.fileBasedMmr.getRoot(blockLeafHeight);
+    }
+    await this.updatePoiMmrRoot(poiBlock.id, mmrRoot);
+    this.nextMmrHeight = poiBlock.id + 1;
+  }
+
+  async updatePoiMmrRoot(id: number, mmrValue: Uint8Array) {
+    const poiBlock = await this.poiRepo.findByPk(id);
+    if (poiBlock.mmrRoot === null) {
+      poiBlock.mmrRoot = mmrValue;
+      await poiBlock.save();
+    } else if (!u8aEq(poiBlock.mmrRoot, mmrValue)) {
+      throw new Error(
+        `Poi block height ${id}, Poi mmr ${u8aToHex(
+          poiBlock.mmrRoot,
+        )} not same as filebased mmr: ${u8aToHex(mmrValue)}`,
+      );
+    }
+  }
+
+  async syncFileBaseFromPoi(): Promise<void> {
+    while (!this.isShutdown) {
+      const poiBlocks = await this.getPoiBlocksByRange(this.nextMmrHeight);
+      if (poiBlocks.length !== 0) {
+        for (const block of poiBlocks) {
+          if (this.nextMmrHeight + 1 < block.id) {
+            for (let i = this.nextMmrHeight + 1; i < block.id; i++) {
+              await this.fileBasedMmr.append(DEFAULT_LEAF);
+              this.nextMmrHeight = i + 1;
+            }
+          }
+          await this.appendMmrNode(block);
+        }
+      } else {
+        await delay(MMR_AWAIT_TIME);
+      }
+    }
+  }
+
+  async getPoiBlocksByRange(startHeight: number): Promise<ProofOfIndex[]> {
+    const poiBlocks = await this.poiRepo.findAll({
+      limit: DEFAULT_FETCH_RANGE,
+      where: { id: { [Op.gte]: startHeight } },
+      order: [['id', 'ASC']],
+    });
+    if (poiBlocks.length !== 0) {
+      // logger.info(`fetched Poi Blocks: [${poiBlocks[0].id},${poiBlocks[poiBlocks.length-1].id}] `)
+      return poiBlocks;
+    } else {
+      return [];
+    }
+  }
+
+  ensureFileBasedMmr(projectMmrPath: string) {
+    let fileBasedDb: FileBasedDb;
+    if (fs.existsSync(projectMmrPath)) {
+      fileBasedDb = FileBasedDb.open(projectMmrPath);
```

Comment:
the flag `as+` already cover the scenario that the file exists. I think the additional scenario need to be tend to is when the folder (not the file) doesn't exists.